src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java

```diff
@@ -151,91 +144,55 @@ public SkyValue compute(SkyKey skyKey, Environment env)
     }
 
-    RepositoryFunction handler;
-    if (rule.getRuleClassObject().isSkylark()) {
-      handler = skylarkHandler;
-    } else {
-      handler = handlers.get(rule.getRuleClass());
-    }
+    RepositoryFunction handler = getHandler(rule);
     if (handler == null) {
       // If we refer to a non repository rule then the repository does not exist.
       return RepositoryDirectoryValue.NO_SUCH_REPOSITORY_VALUE;
     }
 
-    handler.setClientEnvironment(clientEnvironmentSupplier.get());
-
     byte[] ruleSpecificData = handler.getRuleSpecificMarkerData(rule, env);
-    if (ruleSpecificData == null) {
-      return null;
-    }
-    String ruleKey = computeRuleKey(rule, ruleSpecificData);
-    Map<String, String> markerData = new TreeMap<>();
-     if (isFetch.get() && handler.isLocal(rule)) {
-      // Local repositories are fetched regardless of the marker file because the operation is
-      // generally fast and they do not depend on non-local data, so it does not make much sense to
-      // try to cache from across server instances.
-      setupRepositoryRoot(repoRoot);
-      env.getListener().post(new RepositoryFetching(repositoryName.getName(), false));
-      RepositoryDirectoryValue.Builder localRepo =
-          handler.fetch(rule, repoRoot, directories, env, markerData, skyKey);
-      if (localRepo == null) {
-        return null;
-      } else {
-        // We write the marker file for local repository essentially for getting the digest and
-        // injecting it in the RepositoryDirectoryValue.
-        byte[] digest = writeMarkerFile(markerPath, markerData, ruleKey);
-        env.getListener().post(new RepositoryFetching(repositoryName.getName(), true));
-        return localRepo.setDigest(digest).build();
-      }
-    }
-
-    // We check the repository root for existence here, but we can't depend on the FileValue,
-    // because it's possible that we eventually create that directory in which case the FileValue
-    // and the state of the file system would be inconsistent.
-
-    byte[] markerHash = isFilesystemUpToDate(markerPath, rule, ruleKey, handler, env);
     if (env.valuesMissing()) {
       return null;
     }
-    if (DONT_FETCH_UNCONDITIONALLY.equals(fetchUnconditionally)
-        && markerHash != null
-        && repoRoot.exists()) {
-      // Now that we know that it exists and that we should not fetch unconditionally, we can
-      // declare a Skyframe dependency on the repository root.
-      RepositoryFunction.getRepositoryDirectory(repoRoot, env);
-      if (env.valuesMissing()) {
-        return null;
+    DigestWriter digestWriter = new DigestWriter(directories, repositoryName, rule,
+        Preconditions.checkNotNull(ruleSpecificData));
+
+    // Local repositories are fetched regardless of the marker file because the operation is
+    // generally fast and they do not depend on non-local data, so it does not make much sense to
+    // try to cache from across server instances.
```

Comment:
nit: try to cache *them* across server instances

Reply:
Corrected